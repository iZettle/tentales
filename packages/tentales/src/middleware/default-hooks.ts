import { Hook, Config } from "tentales"
import { Log } from "tentales-log"
import { path } from "ramda"
import { bodyParserMiddleware } from "./middlewares/body-parser-middleware"
import { renderMiddleware } from "./middlewares/render-middleware"
import { fourOhFourMiddleware } from "./middlewares/four-oh-four-middleware"
import { errorMiddleware } from "./middlewares/error-middleware"
import { protectServiceRoutesMiddleware } from "./middlewares/protect-service-routes-middelware"

export interface GetDefaultHooksParams {
  log: Log
  config: Config
}

const getProtectServiceRoutesHook = ({
  log,
  config,
}: GetDefaultHooksParams): Hook | undefined => {
  const secret = path(["auth", "serverSecret"], config)

  if (secret) {
    return ["protectServiceRoutesMiddleware", [protectServiceRoutesMiddleware]]
  } else {
    log.warn(
      "Auth secret not set; this will disable authentication checks for all services.",
    )
  }
}

const DEFAULT_HOOKS_PUBLIC = ({ log, config }: GetDefaultHooksParams): Hook[] =>
  [
    ["errorMiddleware", [errorMiddleware]],
    ["bodyParserMiddleware", [bodyParserMiddleware]],
    ["renderMiddleware", [renderMiddleware]],
    getProtectServiceRoutesHook({ log, config }),
    ["fourOhFourMiddleware", [fourOhFourMiddleware]],
  ].filter(Boolean) as Hook[]

const DEFAULT_HOOKS_SERVICE = ({
  log,
  config,
}: GetDefaultHooksParams): Hook[] =>
  [
    ["errorMiddleware", [errorMiddleware]],
    ["bodyParserMiddleware", [bodyParserMiddleware]],
    getProtectServiceRoutesHook({ log, config }),
  ].filter(Boolean) as Hook[]

export function getDefaultHooks({
  log,
  config,
}: GetDefaultHooksParams): Hook[] {
  return config.public
    ? DEFAULT_HOOKS_PUBLIC({ log, config })
    : DEFAULT_HOOKS_SERVICE({ log, config })
}
