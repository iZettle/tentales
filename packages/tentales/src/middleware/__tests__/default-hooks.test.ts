import { getDefaultHooks } from "../default-hooks"

test("It gets public default hooks", async () => {
  const resp = getDefaultHooks({ isPublic: true })
  expect(resp).toMatchSnapshot()
})

test("It gets private default hooks", async () => {
  const resp = getDefaultHooks({ isPublic: false })
  expect(resp).toMatchSnapshot()
})
