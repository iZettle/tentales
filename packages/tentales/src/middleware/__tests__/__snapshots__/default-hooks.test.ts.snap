// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It gets private default hooks 1`] = `
Array [
  Array [
    "errorMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "bodyParserMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "protectServiceRoutesMiddleware",
    Array [
      [Function],
    ],
  ],
]
`;

exports[`It gets public default hooks 1`] = `
Array [
  Array [
    "errorMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "bodyParserMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "renderMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "protectServiceRoutesMiddleware",
    Array [
      [Function],
    ],
  ],
  Array [
    "fourOhFourMiddleware",
    Array [
      [Function],
    ],
  ],
]
`;
